### PR TITLE
[14.0][MIG] contract_sale_generation

### PR DESCRIPTION
**Pull request by rousseldenis**
Supersedes #661 

New features:
- [x] Transmit analytic account from contract to sale order
- [x] Needs #741

### CODE REVIEW

**Comment by rousseldenis** [comment]
@pedrobaeza This one is ready too

File: contract_sale_generation/models/abstract_contract.py [hunk 1]
```diff
@@ -2,16 +2,16 @@
...
-        string="Type",
-        default="invoice",
-        required=True,
-    )
     sale_autoconfirm = fields.Boolean(string="Sale Autoconfirm")
+
+    @api.model
+    def _selection_generation_type(self):
+        res = super()._selection_generation_type()
+        res.append(("sale", "Sale"))
```
Comment (pedrobaeza) [inline contract_sale_generation/models/abstract_contract.py:16]: Is this going to be translatable? Or do you need to add `_()`?

File: contract_sale_generation/models/contract.py [hunk 2]
```diff
@@ -118,42 +113,22 @@ def _prepare_recurring_sales_values(self, date_ref=False):
...
-        invoices = self.env["account.invoice"]
-        # Invoice by companies, so assignation emails get correct context
-        companies_to_invoice = self.read_group(domain, ["company_id"], ["company_id"])
-        for row in companies_to_invoice:
-            contracts_to_invoice = self.search(row["__domain"]).with_context(
-                allowed_company_ids=[row["company_id"][0]]
-            )
-            invoices |= contracts_to_invoice._recurring_create_invoice(date_ref)
-        return invoices
+    def cron_recurring_create_sale(self, date_ref=None):
```
Comment (pedrobaeza) [inline contract_sale_generation/models/contract.py:133]: Should we have a different cron for this generation or use the same cron for all?

File: contract_sale_generation/models/contract_line.py [hunk 3]
```diff
@@ -1,41 +1,41 @@
...
+            self.last_date_invoiced, self.recurring_next_date
+        )
+        sale_line_vals = self._prepare_sale_line_vals(dates, order_id)
+
         order_line = (
             self.env["sale.order.line"]
-            .with_context(
-                force_company=self.contract_id.company_id.id,
-            )
+            .with_company(self.contract_id.company_id.id)
```
Comment (pedrobaeza) [inline contract_sale_generation/models/contract_line.py:32]: `with_company` expects a recordset, isn't it?

File: contract_sale_generation/tests/test_contract_sale.py [hunk 4]
```diff
@@ -119,8 +141,7 @@ def test_check_discount(self):
     def test_contract(self):
         recurring_next_date = to_date("2020-02-15")
         self.assertAlmostEqual(self.contract_line.price_subtotal, 50.0)
-        res = self.contract_line._onchange_product_id()
-        self.assertIn("uom_id", res["domain"])
+        self.contract_line._onchange_product_id()
```
Comment (pedrobaeza) [inline contract_sale_generation/tests/test_contract_sale.py]: Do you still need to call these onchanges?

File: contract_sale_generation/models/contract.py (hunk 2 above)
Comment (rousseldenis) [reply contract_sale_generation/models/contract.py:133]: If you want to have different time slots for the cron, this is mandatory

File: contract_sale_generation/models/contract_line.py (hunk 3 above)
Comment (rousseldenis) [reply contract_sale_generation/models/contract_line.py:32]: That does not have importance:
image

File: contract_sale_generation/models/abstract_contract.py (hunk 1 above)
Comment (rousseldenis) [reply contract_sale_generation/models/abstract_contract.py:16]: The Selection field behaviour does not need it

File: contract_sale_generation/models/contract.py (hunk 2 above)
Comment (rousseldenis) [reply contract_sale_generation/models/contract.py:133]: And this exists in previous version. Don't want to break

**Review by pedrobaeza — APPROVED** [review]

**Comment by rousseldenis** [comment]
@pedrobaeza Done

**Comment by rousseldenis** [comment]
@pedrobaeza green

**Review by pedrobaeza — APPROVED** [review]
Can you get another review for merging it?

**Comment by rousseldenis** [comment]
@pedrobaeza Split tests for better inheritance and cleaned code.

@djulien19 Could you review?

**Review by sanrav — APPROVED** [review]
Functional testing ok

**Pull request merged**